crates/server/src/access.rs

```diff
@@ -0,0 +1,182 @@
+use std::net::IpAddr;
+
+use hickory_proto::error::{ProtoError, ProtoErrorKind};
+use ipnet::{IpNet, Ipv4Net, Ipv6Net};
+use prefix_trie::{Prefix, PrefixSet};
+
+/// Type to evaluate access from a source address for accessing the server.
+///
+/// Allowed networks will override denied networks, i.e. if a network is allowed, the deny rules will not be evaluated.
+/// Allowed networks are processed in the context of denied networks, that is, if there are no denied networks, then
+///   the allowed list will effectively deny access to anything that's not in the allowed list. On the other hand, if
+///   denied networks are specified, then allowed networks will only apply if the deny rule matched, but otherwise the
+///   address will be allowed.
+#[derive(Default)]
+pub(crate) struct Access {
+    deny_ipv4: Option<PrefixSet<Ipv4Net>>,
+    deny_ipv6: Option<PrefixSet<Ipv6Net>>,
+    allow_ipv4: Option<PrefixSet<Ipv4Net>>,
+    allow_ipv6: Option<PrefixSet<Ipv6Net>>,
```

Comment:
What's the point of wrapping all of these in an `Option`? I think things would be simpler if that wasn't the case, and you could instead evaluate if the `PrefixSet` is empty if necessary (it seems surprising that `PrefixSet` doesn't have an `is_empty()` method... but I suggest that this could be simply worked around using something like `iter().count() == 0`).

Reply:
Yeah, PrefixSet doesn't have an is_empty for some reason. I just checked the conversion for the iter impl, it looks ok. I'll see about getting them to add an is_empty.